src/videojs-hls.js

```diff
@@ -906,9 +915,39 @@ videojs.HlsHandler.prototype.fillBuffer = function(mediaIndex) {
     // the state of the buffer before a segment is appended will be
     // stored here so that the actual segment duration can be
     // determined after it has been appended
-    buffered: null
+    buffered: null,
+    // The target timestampOffset for this segment when we append it
+    // to the source buffer
+    timestampOffset: null
   };
 
+  if (mediaIndex > 0) {
+    segmentTimestampOffset = videojs.Hls.Playlist.duration(segmentInfo.playlist,
+      segmentInfo.playlist.mediaSequence + mediaIndex) + this.playlists.expired_;
+  }
+
+  // If we have seeked into a non-buffered time-range, remove all buffered
+  // time-ranges because they could have been incorrectly placed originally
+  if (this.tech_.seeking() && outsideBufferedRanges) {
+    // If there are discontinuities in the playlist, we can't be sure of anything
+    // related to time so we reset the timestamp offset and start appending data
+    // anew on every seek
+    if (segmentInfo.playlist.discontinuityStarts.length) {
+      // Now that the forward buffer is clear, we have to set timestamp offset to
+      // the start of the buffered region
+      segmentInfo.timestampOffset = segmentTimestampOffset;
+    }
+  } else if (segment.discontinuity && currentBuffered.length) {
+    // If we aren't seeking and are crossing a discontinuity, we should set
+    // timestampOffset for new segments to be appended the end of the current
+    // buffered time-range
+    segmentInfo.timestampOffset = currentBuffered.end(0);
```

Comment:
As we discussed, I feel like this whole section must be revisited at some point in the near future but I am hesitant to make changes until we have the first official release of MSE-HLS out the door, so to speak.